### PR TITLE
Testing that batch_encode_plus is the same as encode_plus

### LINKED ISSUE #2654: Add keyword arguments to batch_encode_plus() to match encode_plus()
## 🚀Consistent Keyword arguments for batch_encode_plus() to match encode_plus()
 Currently, features such as `return_special_tokens_mask` that are available for the `encode_plus()` method are not available for `batch_encode_plus()`. It would be nice if all keyword arguments worked in a similar fashion.

## Motivation
`batch_encode_plus()` is extremely useful to seamlessly tokenize batches, however, the lack of features forces to fall back to encode_plus() and an _uglier_ implementation.

### LINKED ISSUE #2658: Inconsistent values returned by batch_encode_plus and encode_plus
## 🐛 Bug

Model I am using (Bert, XLNet....): Bert

Language I am using the model on (English, Chinese....): bert-base-uncased

The problem arise when using:
* batch_encode_plus and encode_plus with pad_to_max_length & max_length

## To Reproduce

Minimal exemple :

I compare tokens created via batch or not, and they are different for **masks** and **types**
```
pretrained = 'bert-base-uncased'
tokenizer = FixedAutoTokenizer.from_pretrained(pretrained)
model = AutoModel.from_pretrained(pretrained)
text = "My"
text1 = "features are ok"

mylist = list()
mylist.append(text)
mylist.append(text1)

#####################################################################################################
batch_encoding = tokenizer.batch_encode_plus(mylist,
                                             return_tensors='pt',
                                             add_special_tokens=False)

######################################################################################################
text_encoding = tokenizer.encode_plus(text,
                                      return_tensors='pt',
                                      add_special_tokens=False,
                                      max_length=3,
                                      pad_to_max_length=True)

print("\n--Batch Encoding \n")
print(batch_encoding['input_ids'])
print(batch_encoding['token_type_ids'])
print(batch_encoding['attention_mask'])
print("\n--One-by-one encoding\n")
print(text_encoding['input_ids'])
print(text_encoding['token_type_ids'])
print(text_encoding['attention_mask'])
```

It gives 

```
--Batch Encoding 

tensor([[2026,    0,    0],
        [2838, 2024, 7929]])
tensor([[0, 1, 1],
        [0, 0, 0]])
tensor([[1, 1, 1],
        [1, 1, 1]])

--One-by-one encoding

tensor([[2026,    0,    0]])
tensor([[0, 0, 0]])
tensor([[1, 0, 0]])
```

## Expected behavior

It should return the same value, like following 

```
--Batch Encoding 

tensor([[2026,    0,    0],
        [2838, 2024, 7929]])
tensor([[0, 0, 0],
        [0, 0, 0]])
tensor([[1, 0, 0],
        [1, 1, 1]])

--One-by-one encoding

tensor([[2026,    0,    0]])
tensor([[0, 0, 0]])
tensor([[1, 0, 0]])
```

## Environment

* OS: MacOs Mojave (reproduced also on Ubuntu)
* Python version: 3.7
* PyTorch version: torch==1.3.1
* PyTorch Transformers version (or branch): transformers==2.3.0
* Using GPU ? no

### LINKED ISSUE #2960: Python Tokenizer batch_encode_plus doesn't pad input if asked to do so.
```python
input_p = tokenizer_p.batch_encode_plus(
    ["This is a simple input 1", "This is a simple input 2"], 
    max_length=15, 
    pad_to_max_length=True
)
```

Output is not padded

### PR DESCRIPTION
**Pull request by LysandreJik**
Spoiler alert: it wasn't.

closes #2960 
closes #2658 
closes #2654

### CODE REVIEW

File: src/transformers/tokenization_utils.py [hunk 1]
```diff
@@ -1096,42 +1140,53 @@ def batch_encode_plus(
...
+                        else:
+                            raise ValueError(
+                                "The sequences building the batch are not of the same size, no tensor "
+                                "can be built. Set `pad_to_max_length=True` to pad the smaller sequences"
+                                "up to the larger sequence's length."
+                            )
                 elif return_tensors == "pt" and is_torch_available():
-                    batch_outputs[key] = torch.tensor(padded_value)
+                    try:
+                        batch_outputs[key] = torch.tensor(value)
```
Comment (sshleifer) [inline src/transformers/tokenization_utils.py]: (nittiest ever) same as torch.Tensor (upper case?) I assume?

File: src/transformers/tokenization_utils.py [hunk 2]
```diff
@@ -1096,42 +1140,53 @@ def batch_encode_plus(
...
+                    except ValueError:
+                        raise ValueError(
+                            "The sequences building the batch are not of the same size, no tensor can be "
+                            "built. Set `pad_to_max_length=True` to pad the smaller sequences up to the"
+                            "larger sequence's length."
+                        )
+                    except RuntimeError:
+                        if None in [item for sequence in value for item in sequence]:
+                            raise ValueError(
```
Comment (sshleifer) [inline src/transformers/tokenization_utils.py]: good msg!

File: src/transformers/tokenization_utils.py [hunk 3]
```diff
@@ -1096,42 +1140,53 @@ def batch_encode_plus(
...
-                    ]
-
                 if return_tensors == "tf" and is_tf_available():
-                    batch_outputs[key] = tf.constant(padded_value)
+                    try:
+                        batch_outputs[key] = tf.constant(value)
+                    except ValueError:
+                        if None in [item for sequence in value for item in sequence]:
+                            raise ValueError(
+                                "No padding token is set for this model, therefore no batch can be made"
```
Comment (sshleifer) [inline src/transformers/tokenization_utils.py]: NO_PAD_TOKEN_MSG = "" as a module level constant will make this smaller.

File: src/transformers/tokenization_utils.py [hunk 4]
```diff
@@ -1067,21 +1089,43 @@ def batch_encode_plus(
...
             if isinstance(ids_or_pair_ids, (list, tuple)):
                 assert len(ids_or_pair_ids) == 2
                 ids, pair_ids = ids_or_pair_ids
             else:
                 ids, pair_ids = ids_or_pair_ids, None
-            outputs = self.encode_plus(
-                ids,
-                pair_ids,
-                add_special_tokens=add_special_tokens,
+
```
Comment (sshleifer) [inline src/transformers/tokenization_utils.py]: so we stopped calling encode plus?

File: src/transformers/tokenization_utils.py [hunk 5]
```diff
@@ -1058,6 +1067,19 @@ def batch_encode_plus(
             **kwargs: passed to the `self.tokenize()` method
```
Comment (sshleifer) [inline src/transformers/tokenization_utils.py]: mention that returns dict?

File: src/transformers/tokenization_utils.py (hunk 4 above)
Comment (sshleifer) [reply src/transformers/tokenization_utils.py]: also cant you call get_input_ids inside the loop, since it takes a tuple?
and then append without ever splitting first and second ids?

File: src/transformers/tokenization_utils.py [hunk 6]
```diff
@@ -1067,21 +1089,43 @@ def batch_encode_plus(
...
+                if second_ids is not None:
+                    return len(first_ids) + len(second_ids) + self.max_len - self.max_len_sentences_pair
+                else:
+                    return len(first_ids) + self.max_len - self.max_len_single_sentence
+
+            max_length = max([total_sequence_length(ids) for ids in input_ids])
+
+        batch_outputs = {}
+        for first_ids, second_ids in input_ids:
+            outputs = self.prepare_for_model(
```
Comment (sshleifer) [inline src/transformers/tokenization_utils.py]: comment about what prepare_for_model does would be nice.

**Review by sshleifer — COMMENTED** [review]
Seems like you are moving work from encode_plus to prepare_for_model? (Is that right?)
Does that merit a name change?

File: src/transformers/tokenization_utils.py (hunk 1 above)
Comment (mfuntowicz) [reply src/transformers/tokenization_utils.py]: torch.tensor is a factory method which will dispatch creating the actual torch.Tensor implementation  (torch.cuda.DoubleTensor, torch.FloatTensor, etc.) to right type of tensor (float, double, long, byte, bool, etc.) on the right device (device=...) and with the current named tensor parameters / alignment in memory

**Review by mfuntowicz — APPROVED** [review]

File: src/transformers/tokenization_utils.py (hunk 3 above)
Comment (LysandreJik) [reply src/transformers/tokenization_utils.py]: You're right, that would be better indeed!

File: src/transformers/tokenization_utils.py (hunk 4 above)
Comment (LysandreJik) [reply src/transformers/tokenization_utils.py]: We stopped calling `encode_plus`, which does the following -> 

- Gets strings and converts them to ids *or* gets tokens and converts them to ids *or* gets ids and does nothing with them
- Passes the resulting ids to the `prepare_for_model` method.

`batch_encode_plus` does the same thing, just with an array instead. I changed from calling `encode_plus` to `prepare_for_model` as I need to know the sequence that, once encoded, has the maximum length in that batch as well as that length. This will be used when setting `pad_to_max_length=True`, without a `max_length`, so that it pads the smaller sequences up to the longer sequence's length (useful for creating tensors out of that batch). 

I'm not sure I understand your second comment, though. Do you mean call `get_input_ids` inside the if statement?

File: src/transformers/tokenization_utils.py (hunk 6 above)
Comment (LysandreJik) [reply src/transformers/tokenization_utils.py]: Correct!

File: src/transformers/tokenization_utils.py [hunk 7]
```diff
@@ -1067,21 +1122,46 @@ def batch_encode_plus(
...
+            input_ids.append((first_ids, second_ids))
+
+        if max_length is None and pad_to_max_length:
+
+            def total_sequence_length(input_pairs):
+                first_ids, second_ids = input_pairs
+                if second_ids is not None:
+                    return len(first_ids) + len(second_ids) + self.max_len - self.max_len_sentences_pair
+                else:
+                    return len(first_ids) + self.max_len - self.max_len_single_sentence
```
Comment (LysandreJik) [inline src/transformers/tokenization_utils.py]: Could be simplified by using `num_added_tokens()`

File: src/transformers/tokenization_utils.py (hunk 4 above)
Comment (sshleifer) [reply src/transformers/tokenization_utils.py]: Tap me, I can explain first comment.

Re padding: does https://pytorch.org/docs/stable/nn.html#torch.nn.utils.rnn.pad_sequence help? `torch.nn.utils.rnn.pad_sequence(sequences, batch_first=False, padding_value=0)`

Tap me if you want to discuss the second comment!

**Review by sshleifer — APPROVED** [review]

File: src/transformers/tokenization_utils.py (hunk 4 above)
Comment (julien-c) [reply src/transformers/tokenization_utils.py]: Agree that we could probably use pad_sequence if it makes the code simpler (I use it in `run_language_modeling`)

**Review by julien-c — APPROVED** [review]
LGTM on a very superficial glance
